### ISSUE TITLE
Connection to broker lost. Trying to re-establish the connection: OSError: [Errno 9] Bad file descriptor

### ISSUE THREAD

**Issue by cajbecu** [celery/celery#4457]
## Checklist

- [ ] I have included the output of ``celery -A proj report`` in the issue.
      (if you are not able to do this, then at least specify the Celery
       version affected).
- [x] I have verified that the issue exists against the `master` branch of Celery.

## Software
celery==4.1.0
kombu==4.1.0
amqp==2.2.2
Python 3.6.1
broker: rabbitmq 3.6.14
result backend: redis

## Steps to reproduce
1. celery -A proj worker -Q Q1 --autoscale=10,1 -Ofair --without-gossip --without-mingle --heartbeat-interval=60 -n Q1
2. celery lost connection to broker
3. after restarting affected worker the connection is successfully re-established and the worker starts processing tasks

## Expected behavior
celery should re-establish connection to broker

## Actual behavior
celery tries to re-establish connection to broker but fails with this error message (which is repeated every second) until manually restarted:
```
[user] celery.worker.consumer.consumer WARNING 2017-12-18 00:38:27,078 consumer: 
Connection to broker lost. Trying to re-establish the connection...
Traceback (most recent call last):
  File "/home/user/.envs/user/lib/python3.6/site-packages/celery/worker/consumer/consumer.py", line 320, in start
    blueprint.start(self)
  File "/home/user/.envs/user/lib/python3.6/site-packages/celery/bootsteps.py", line 119, in start
    step.start(parent)
  File "/home/user/.envs/user/lib/python3.6/site-packages/celery/worker/consumer/consumer.py", line 596, in start
    c.loop(*c.loop_args())
  File "/home/user/.envs/user/lib/python3.6/site-packages/celery/worker/loops.py", line 47, in asynloop
    obj.controller.register_with_event_loop(hub)
  File "/home/user/.envs/user/lib/python3.6/site-packages/celery/worker/worker.py", line 217, in register_with_event_loop
    description='hub.register',
  File "/home/user/.envs/user/lib/python3.6/site-packages/celery/bootsteps.py", line 151, in send_all
    fun(parent, *args)
  File "/home/user/.envs/user/lib/python3.6/site-packages/celery/worker/components.py", line 178, in register_with_event_loop
    w.pool.register_with_event_loop(hub)
  File "/home/user/.envs/user/lib/python3.6/site-packages/celery/concurrency/prefork.py", line 134, in register_with_event_loop
    return reg(loop)
  File "/home/user/.envs/user/lib/python3.6/site-packages/celery/concurrency/asynpool.py", line 476, in register_with_event_loop
    for fd in self._fileno_to_outq]
  File "/home/user/.envs/user/lib/python3.6/site-packages/celery/concurrency/asynpool.py", line 476, in <listcomp>
    for fd in self._fileno_to_outq]
  File "/home/user/.envs/user/lib/python3.6/site-packages/kombu/async/hub.py", line 207, in add_reader
    return self.add(fds, callback, READ | ERR, args)
  File "/home/user/.envs/user/lib/python3.6/site-packages/kombu/async/hub.py", line 158, in add
    self.poller.register(fd, flags)
  File "/home/user/.envs/user/lib/python3.6/site-packages/kombu/utils/eventio.py", line 67, in register
    self._epoll.register(fd, events)
OSError: [Errno 9] Bad file descriptor
```

**Comment by ryanpetrello** [comment]
I work on an open source project that uses celery and is encountering this bug also.  One thing we've discovered is that you can cause celery to break out of this failure by sending it `SIGHUP`.

**Comment by auvipy** [comment]
could you both try celery 4.2rc4 and see if it still persist?

**Comment by wind2008hxy** [comment]
same issues

**Comment by auvipy** [comment]
are you facing this with celery 4.2?

**Comment by wind2008hxy** [comment]
I upgrade to celery 4.2.0,still have the same issue

**Comment by wind2008hxy** [comment]
old celery verison is 3.1.23, I try  to upgrade celery 3.x to 4.x. 

broker&backend is redis

old version 3.1.23 work well, but where use 4.2.0, it will `Connection to broker lost. Trying to re-establish the connection`

```
   'CELERY_FORCE_ROOT': False,
    'CELERY_TASK_SERIALIZER': 'pickle',
    'CELERY_RESULT_SERIALIZER': 'pickle',
    'CELERY_ACCEPT_CONTENT': ['pickle', 'json'],
    'CELERY_TASK_PROTOCOL': 1,
    'CELERY_BROKER_URL': 'redis://127.0.0.1:6379/0',
    'CELERY_RESULT_BACKEND': "redis://127.0.0.1:6379/0",
    'CELERY_REDIS_MAX_CONNECTIONS': 2048,
    'CELERY_TASK_RESULT_EXPIRES': 86400 * 7,
    'CELERY_TRACK_STARTED': True,
    'CELERY_ENABLE_UTC': False,
    'CELERY_TIMEZONE': 'Asia/Shanghai',
    'CELERY_IGNORE_RESULT': True,
    'CELERYD_MAX_TASKS_PER_CHILD': 500,
    'CELERYD_TASK_SOFT_TIME_LIMIT': 60 * 5,
    'CELERYD_PREFETCH_MULTIPLIER': 20,
```

**Comment by tonal** [comment]
see also #1115, #1290

**Comment by tonal** [comment]
Same to RabbitMQ:
```
[2018-10-10 12:25:27,607: WARNING/MainProcess] consumer: Connection to broker lost. Trying to re-establish the connection...
Traceback (most recent call last):
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/worker/consumer/consumer.py", line 317, in start
    blueprint.start(self)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/bootsteps.py", line 119, in start
    step.start(parent)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/worker/consumer/consumer.py", line 593, in start
    c.loop(*c.loop_args())
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/worker/loops.py", line 91, in asynloop
    next(loop)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/kombu/asynchronous/hub.py", line 354, in create_loop
    cb(*cbargs)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/kombu/transport/base.py", line 236, in on_readable
    reader(loop)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/kombu/transport/base.py", line 218, in _read
    drain_events(timeout=0)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/amqp/connection.py", line 491, in drain_events
    while not self.blocking_read(timeout):
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/amqp/connection.py", line 497, in blocking_read
    return self.on_inbound_frame(frame)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/amqp/method_framing.py", line 55, in on_frame
    callback(channel, method_sig, buf, None)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/amqp/connection.py", line 501, in on_inbound_method
    method_sig, payload, content,
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/amqp/abstract_channel.py", line 128, in dispatch_method
    listener(*args)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/amqp/connection.py", line 621, in _on_close
    self._x_close_ok()
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/amqp/connection.py", line 636, in _x_close_ok
    self.send_method(spec.Connection.CloseOk, callback=self._on_close_ok)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/amqp/abstract_channel.py", line 50, in send_method
    conn.frame_writer(1, self.channel_id, sig, args, content)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/amqp/method_framing.py", line 166, in write_frame
    write(view[:offset])
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/amqp/transport.py", line 275, in write
    self._write(s)
ConnectionResetError: [Errno 104] Connection reset by peer
[2018-10-10 12:25:27,863: ERROR/MainProcess] consumer: Cannot connect to amqp://remains:**@127.0.0.1:5672//: [Errno 111] Connection refused.
Trying again in 2.00 seconds...

[2018-10-10 12:25:29,878: ERROR/MainProcess] consumer: Cannot connect to amqp://remains:**@127.0.0.1:5672//: [Errno 111] Connection refused.
Trying again in 4.00 seconds...

[2018-10-10 12:25:33,893: ERROR/MainProcess] consumer: Cannot connect to amqp://remains:**@127.0.0.1:5672//: [Errno 111] Connection refused.
Trying again in 6.00 seconds...

[2018-10-10 12:25:39,935: INFO/MainProcess] Connected to amqp://remains:**@127.0.0.1:5672//
[2018-10-10 12:25:39,953: INFO/MainProcess] mingle: searching for neighbors
[2018-10-10 12:25:39,958: WARNING/MainProcess] consumer: Connection to broker lost. Trying to re-establish the connection...
Traceback (most recent call last):
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/worker/consumer/consumer.py", line 317, in start
    blueprint.start(self)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/bootsteps.py", line 119, in start
    step.start(parent)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/worker/consumer/mingle.py", line 40, in start
    self.sync(c)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/worker/consumer/mingle.py", line 44, in sync
    replies = self.send_hello(c)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/worker/consumer/mingle.py", line 57, in send_hello
    replies = inspect.hello(c.hostname, our_revoked._data) or {}
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/app/control.py", line 143, in hello
    return self._request('hello', from_node=from_node, revoked=revoked)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/app/control.py", line 95, in _request
    timeout=self.timeout, reply=True,
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/app/control.py", line 454, in broadcast
    limit, callback, channel=channel,
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/kombu/pidbox.py", line 315, in _broadcast
    serializer=serializer)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/kombu/pidbox.py", line 290, in _publish
    serializer=serializer,
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/kombu/messaging.py", line 181, in publish
    exchange_name, declare,
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/kombu/messaging.py", line 203, in _publish
    mandatory=mandatory, immediate=immediate,
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/amqp/channel.py", line 1732, in _basic_publish
    (0, exchange, routing_key, mandatory, immediate), msg
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/amqp/abstract_channel.py", line 50, in send_method
    conn.frame_writer(1, self.channel_id, sig, args, content)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/amqp/method_framing.py", line 166, in write_frame
    write(view[:offset])
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/amqp/transport.py", line 275, in write
    self._write(s)
ConnectionResetError: [Errno 104] Connection reset by peer
[2018-10-10 12:25:40,032: INFO/MainProcess] Connected to amqp://remains:**@127.0.0.1:5672//
[2018-10-10 12:25:40,044: INFO/MainProcess] mingle: searching for neighbors
[2018-10-10 12:25:41,062: INFO/MainProcess] mingle: all alone
[2018-10-10 12:25:41,085: WARNING/MainProcess] consumer: Connection to broker lost. Trying to re-establish the connection...
Traceback (most recent call last):
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/worker/consumer/consumer.py", line 317, in start
    blueprint.start(self)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/bootsteps.py", line 119, in start
    step.start(parent)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/worker/consumer/consumer.py", line 593, in start
    c.loop(*c.loop_args())
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/worker/loops.py", line 50, in asynloop
    obj.controller.register_with_event_loop(hub)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/worker/worker.py", line 219, in register_with_event_loop
    description='hub.register',
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/bootsteps.py", line 151, in send_all
    fun(parent, *args)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/worker/components.py", line 179, in register_with_event_loop
    w.pool.register_with_event_loop(hub)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/concurrency/prefork.py", line 134, in register_with_event_loop
    return reg(loop)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/concurrency/asynpool.py", line 486, in register_with_event_loop
    for fd in self._fileno_to_outq]
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/celery/concurrency/asynpool.py", line 486, in <listcomp>
    for fd in self._fileno_to_outq]
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/kombu/asynchronous/hub.py", line 207, in add_reader
    return self.add(fds, callback, READ | ERR, args)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/kombu/asynchronous/hub.py", line 158, in add
    self.poller.register(fd, flags)
  File "/opt/pyenv/versions/3.6.6/envs/remains/lib/python3.6/site-packages/kombu/utils/eventio.py", line 67, in register
    self._epoll.register(fd, events)
OSError: [Errno 9] Bad file descriptor
```

**Comment by rabbit8ge** [comment]
FYI. In my case, I used chain to run task2 after task1 finished. When the return value of task1 is too large (~100M) and large number of tasks are executing at the same time, the problem occurs. Set and get return value to redis by myself resolved the problem.

**Comment by auvipy** [comment]
@rabbit8ge could you come up with a pr with an appropriate fix?

**Comment by sudhishvnair** [comment]
I am also facing the same issue with Celery and RabbitMq.

Celery Version         : v4.3.0rc2
RabbitMq Version    : 3.6.10, Erlang 20.2.2
OS                           : Ubuntu
Python                     : 3.6

Error Trace
------------------------------------------------
[2019-03-26 10:24:06,540: WARNING/MainProcess] consumer: Connection to broker lost. Trying to re-establish the connection...
Traceback (most recent call last):
File "/usr/local/lib/python3.6/dist-packages/celery/worker/consumer/consumer.py", line 316, in start
blueprint.start(self)
File "/usr/local/lib/python3.6/dist-packages/celery/bootsteps.py", line 119, in start
step.start(parent)
File "/usr/local/lib/python3.6/dist-packages/celery/worker/consumer/consumer.py", line 592, in start
c.loop(*c.loop_args())
File "/usr/local/lib/python3.6/dist-packages/celery/worker/loops.py", line 50, in asynloop
obj.controller.register_with_event_loop(hub)
File "/usr/local/lib/python3.6/dist-packages/celery/worker/worker.py", line 219, in register_with_event_loop
description='hub.register',
File "/usr/local/lib/python3.6/dist-packages/celery/bootsteps.py", line 151, in send_all
fun(parent, *args)
File "/usr/local/lib/python3.6/dist-packages/celery/worker/components.py", line 179, in register_with_event_loop
w.pool.register_with_event_loop(hub)
File "/usr/local/lib/python3.6/dist-packages/celery/concurrency/prefork.py", line 134, in register_with_event_loop
return reg(loop)
File "/usr/local/lib/python3.6/dist-packages/celery/concurrency/asynpool.py", line 475, in register_with_event_loop
for fd in self._fileno_to_outq]
File "/usr/local/lib/python3.6/dist-packages/celery/concurrency/asynpool.py", line 475, in <listcomp>
for fd in self._fileno_to_outq]
File "/usr/local/lib/python3.6/dist-packages/kombu/asynchronous/hub.py", line 207, in add_reader
return self.add(fds, callback, READ | ERR, args)
File "/usr/local/lib/python3.6/dist-packages/kombu/asynchronous/hub.py", line 158, in add
self.poller.register(fd, flags)
File "/usr/local/lib/python3.6/dist-packages/kombu/utils/eventio.py", line 67, in register
self._epoll.register(fd, events)
OSError: [Errno 9] Bad file descriptor


Also we could see below error in the log file as well


[2019-03-26 10:24:05,048: WARNING/MainProcess] consumer: Connection to broker lost. Trying to re-establish the connection...
Traceback (most recent call last):
File "/usr/local/lib/python3.6/dist-packages/celery/worker/consumer/consumer.py", line 316, in start
blueprint.start(self)
File "/usr/local/lib/python3.6/dist-packages/celery/bootsteps.py", line 119, in start
step.start(parent)
File "/usr/local/lib/python3.6/dist-packages/celery/worker/consumer/mingle.py", line 40, in start
self.sync(c)
File "/usr/local/lib/python3.6/dist-packages/celery/worker/consumer/mingle.py", line 44, in sync
replies = self.send_hello(c)
File "/usr/local/lib/python3.6/dist-packages/celery/worker/consumer/mingle.py", line 57, in send_hello
replies = inspect.hello(c.hostname, our_revoked._data) or
{}

File "/usr/local/lib/python3.6/dist-packages/celery/app/control.py", line 132, in hello
return self._request('hello', from_node=from_node, revoked=revoked)
File "/usr/local/lib/python3.6/dist-packages/celery/app/control.py", line 84, in _request
timeout=self.timeout, reply=True,
File "/usr/local/lib/python3.6/dist-packages/celery/app/control.py", line 439, in broadcast
limit, callback, channel=channel,
File "/usr/local/lib/python3.6/dist-packages/kombu/pidbox.py", line 321, in _broadcast
channel=chan)
File "/usr/local/lib/python3.6/dist-packages/kombu/pidbox.py", line 360, in _collect
self.connection.drain_events(timeout=timeout)
File "/usr/local/lib/python3.6/dist-packages/kombu/connection.py", line 301, in drain_events
return self.transport.drain_events(self.connection, **kwargs)
File "/usr/local/lib/python3.6/dist-packages/kombu/transport/pyamqp.py", line 103, in drain_events
return connection.drain_events(**kwargs)
File "/usr/local/lib/python3.6/dist-packages/amqp/connection.py", line 491, in drain_events
while not self.blocking_read(timeout):
File "/usr/local/lib/python3.6/dist-packages/amqp/connection.py", line 496, in blocking_read
frame = self.transport.read_frame()
File "/usr/local/lib/python3.6/dist-packages/amqp/transport.py", line 243, in read_frame
frame_header = read(7, True)
File "/usr/local/lib/python3.6/dist-packages/amqp/transport.py", line 418, in _read
s = recv(n - len(rbuf))
ConnectionResetError: [Errno 104] Connection reset by peer

**Comment by sudhishvnair** [comment]
@auvipy have you got a chance to look at my comments?

**Comment by auvipy** [comment]
this also happens in celery 4.3 right?

**Comment by sudhishvnair** [comment]
Yet to check in 4.3 release. I will update it as soon as I try it.

**Comment by sudhishvnair** [comment]
@auvipy This happens in 4.3 release as well. Connection drop has increase dramatically.

**Comment by matteius** [comment]
This seems to not be specific to the Redis broker .... we have this issue too and use Rabbit MQ when we experience it.  One thing to note, both us and the OP start celery worker similarly with --without-gossip --without-mingle --heartbeat-interval=60 however we never experienced the issue until the release where we added an --autoscale setting such as --autoscale=3,1 ... so perhaps looking at how that code works may be a good starting point to a PR.

**Comment by matteius** [comment]
I spent a couple hours this morning trying to read through the code more and while I have nothing totally conclusive yet based on the stack traces occuring in the celery AsynPool class, looking in there I can see that it sets there different fileno to process mapping dictionaries in the init such as the one named _fileno_to_outq from the stack trace.  From reading through the code it looks like this mapping gets built up when the workers start, but for autoscaling I don't see where things are actually removed from this mapping when a process exits.  Perhaps the bug is around this class?  https://github.com/celery/celery/blob/master/celery/concurrency/asynpool.py#L396

**Comment by auvipy** [comment]
It would be great if you could start a pr with failing test case and related changes which seems to be the root cause

**Comment by auvipy** [comment]
I request everyone to check the open PR

**Comment by bpereto** [comment]
sorry, i'm late to the party. We currently test the proposed fix for the same Issue in celery 3.25. The difference is that in our case an IOError is throwed instead of an OSError with the same location and traceback.

The error could be catched also:
```
except (OSError, IOError):
```

**Comment by matteius** [comment]
@bpereto   I actually modified the code to catch OSError but there will have to be more to the fix -- catching the error alone is not enough, in my logging annotated version of drain_events_until that catches OS Error I find that  in the current flows because it gets stuck calling wait_for the time after the OSError is raised.
`    def drain_events_until(self, p, timeout=None, on_interval=None, wait=None):
        wait = wait or self.result_consumer.drain_events
        time_start = monotonic()

        while 1:
            # Total time spent may exceed a single call to wait()
            if timeout and monotonic() - time_start >= timeout:
                raise socket.timeout()
            try:
                log.info("Calling self.wait_for(p)")
                yield self.wait_for(p, wait, timeout=timeout)
            except IOError:
                log.info("IO ERROR IN drain_events_until")
                pass
            if on_interval:
                log.info("Calling on_interval")
                on_interval()
                log.info("on_interval returned")
            log.info(type(p))
            if p.ready:  # got event on the wanted channel.
                break`

I think the OSError vs IOError may be a py2 vs py3 difference?  In any case for the early versions of 4 (not sure about 3) we did not actually pass in the configured heartbeat parameter to the worker and so it got the timeout of 0 and acted as a nonblocking socket thus working because the socket didn't close after some specified timeout condition (current running theory).
